File: src/orchestrator.ts
import path from "node:path";

import { CheckpointManager } from "./checkpoint.js";
import type { AuditConfig, EntryPoint, Module } from "./config.js";
import { getLogger } from "./logger.js";
import { getInScopeModules } from "./parsing/stage2.js";
import { parseEntryPoints } from "./parsing/stage3.js";
import { listMarkdownFiles } from "./utils.js";
import { runSetup } from "./stages/stage0.js";
import { runStage1 } from "./stages/stage1.js";
import { runStage2 } from "./stages/stage2.js";
import { runStage3 } from "./stages/stage3.js";
import { runStage4 } from "./stages/stage4.js";
import { runStage5 } from "./stages/stage5.js";
import { runStage6 } from "./stages/stage6.js";

const logger = getLogger("orchestrator");

export async function runAudit(config: AuditConfig): Promise<string> {
  const checkpoint = new CheckpointManager(config.outputDir, config.resume);

  if (config.resume) {
    logger.info("Resume mode enabled. Existing output files and markers will be reused.");
  }

  if (!config.skipStages.includes(0)) {
    await runSetup(config);
  }

  let instructionStage2Path: string;
  let instructionStage5Path: string;
  if (!config.skipStages.includes(1)) {
    const stage1 = await runStage1(config, checkpoint);
    instructionStage2Path = stage1.instructionStage2Path;
    instructionStage5Path = stage1.instructionStage5Path;
  } else {
    logger.info("Stage 1 skipped.");
    instructionStage2Path = path.join(config.outputDir, "stage-1-details", "instruction-stage2.md");
    instructionStage5Path = path.join(config.outputDir, "stage-1-details", "instruction-stage5.md");
  }

  let modules: Module[] = [];
  if (!config.skipStages.includes(2)) {
    modules = await runStage2(config, checkpoint, instructionStage2Path);
  } else {
    logger.info("Stage 2 skipped.");
    modules = getInScopeModules(path.join(config.outputDir, "stage-2-scope.md"));
  }

  if (modules.length === 0) {
    throw new Error("Stage 2 produced no in-scope modules.");
  }

  let entryPointMap: Record<string, EntryPoint[]> = {};
  if (!config.skipStages.includes(3)) {
    entryPointMap = await runStage3(modules, config, checkpoint);
  } else {
    logger.info("Stage 3 skipped.");
    const stage3Dir = path.join(config.outputDir, "stage-3-details");
    for (const module of modules) {
      const filePath = path.join(stage3Dir, `${module.id}.md`);
      entryPointMap[module.id] = parseEntryPoints(filePath, module.id);
    }
  }

  let findingFiles: string[] = [];
  if (!config.skipStages.includes(4)) {
    findingFiles = await runStage4(entryPointMap, config, checkpoint);
  } else {
    logger.info("Stage 4 skipped.");
    findingFiles = await listMarkdownFiles(path.join(config.outputDir, "stage-4-details"));
  }

  if (!config.skipStages.includes(5)) {
    await runStage5(findingFiles, config, checkpoint, instructionStage5Path);
  } else {
    logger.info("Stage 5 skipped.");
  }

  let reportPath = "";
  if (!config.skipStages.includes(6)) {
    reportPath = await runStage6(config, checkpoint);
  } else {
    logger.info("Stage 6 skipped.");
  }

  logger.info("Audit complete. Report: %s", reportPath);
  return reportPath;
}

File: code_auditor/orchestrator.py
from __future__ import annotations

import os

from .checkpoint import CheckpointManager
from .config import AnalysisUnit, AuditConfig
from .logger import get_logger
from .parsing.stage2 import parse_au_files
from .stages.stage0 import run_setup
from .stages.stage1 import Stage1Output, run_stage1
from .stages.stage2 import run_stage2
from .stages.stage3 import run_stage3
from .stages.stage4 import run_stage4
from .stages.stage5 import run_stage5
from .stages.stage6 import run_stage6
from .tui import TUIManager
from .utils import _natural_sort_key, list_json_files

logger = get_logger("orchestrator")


async def run_audit(config: AuditConfig, tui: TUIManager | None = None) -> None:
    checkpoint = CheckpointManager(config.output_dir, config.resume)

    if config.resume:
        logger.info("Resume mode enabled. Existing output files and markers will be reused.")

    # Stage 0: setup
    if 0 not in config.skip_stages:
        if tui:
            tui.begin_stage(0, "Setting up output directory")
        await run_setup(config)
        if tui:
            tui.end_stage(0)
    else:
        if tui:
            tui.skip_stage(0)

    # Stage 1: security context research
    stage1_out: Stage1Output | None = None
    if 1 not in config.skip_stages:
        if tui:
            tui.begin_stage(1, "Researching security context")
        stage1_out = await run_stage1(config, checkpoint)
        if tui:
            tui.end_stage(1)
    else:
        if tui:
            tui.skip_stage(1)

    # Resolve directive paths (from stage1 output or default locations)
    details_dir = os.path.join(config.output_dir, "stage1-security-context")
    auditing_focus_path = (
        stage1_out.auditing_focus_path if stage1_out
        else os.path.join(details_dir, "auditing-focus.md")
    )
    vuln_criteria_path = (
        stage1_out.vuln_criteria_path if stage1_out
        else os.path.join(details_dir, "vulnerability-criteria.md")
    )

    # Stage 2: decompose project into analysis units
    analysis_units: list[AnalysisUnit] = []
    if 2 not in config.skip_stages:
        if tui:
            tui.begin_stage(2, "Decomposing codebase")
        analysis_units = await run_stage2(config, checkpoint, auditing_focus_path)
        if tui:
            tui.stage_progress(2, items_done=len(analysis_units), items_total=len(analysis_units),
                               detail=f"{len(analysis_units)} analysis units found")
            tui.end_stage(2)
    else:
        logger.info("Stage 2 skipped. Loading existing analysis units.")
        stage2_dir = os.path.join(config.output_dir, "stage2-analysis-units")
        analysis_units = parse_au_files(stage2_dir)
        if tui:
            tui.skip_stage(2, f"Loaded {len(analysis_units)} existing AUs")

    if not analysis_units and 3 not in config.skip_stages:
        raise RuntimeError("Stage 2 produced no analysis units.")

    # Stage 3: bug discovery per AU
    bug_files: list[str] = []
    if 3 not in config.skip_stages:
        total_aus = len(analysis_units)
        if tui:
            tui.begin_stage(3, f"Discovering bugs across {total_aus} AUs")

        # Wrap the stage3 call with progress tracking
        _original_done = [0]
        _stage3_tui = tui

        class _ProgressProxy:
            """Wraps run_stage3 to emit progress events to the TUI."""
            def __init__(self):
                self.files: list[str] = []

        bug_files = await run_stage3(
            analysis_units, config, checkpoint,
            auditing_focus_path, vuln_criteria_path,
        )
        if tui:
            tui.stage_progress(3, items_done=total_aus, items_total=total_aus,
                               detail=f"{len(bug_files)} findings")
            tui.end_stage(3)
    else:
        logger.info("Stage 3 skipped.")
        bug_files = list_json_files(os.path.join(config.output_dir, "stage3-findings"))
        if tui:
            tui.skip_stage(3, f"Loaded {len(bug_files)} existing findings")

    # Stage 4: evaluate findings
    vuln_files: list[str] = []
    if 4 not in config.skip_stages:
        if tui:
            tui.begin_stage(4, f"Evaluating {len(bug_files)} findings")
        vuln_files = await run_stage4(bug_files, config, checkpoint, vuln_criteria_path)
        if tui:
            tui.stage_progress(4, items_done=len(vuln_files), items_total=len(bug_files),
                               detail=f"{len(vuln_files)} confirmed vulnerabilities")
            tui.end_stage(4)
    else:
        logger.info("Stage 4 skipped.")
        stage4_dir = os.path.join(config.output_dir, "stage4-vulnerabilities")
        vuln_files = [f for f in list_json_files(stage4_dir) if "_pending" not in f]
        if tui:
            tui.skip_stage(4, f"Loaded {len(vuln_files)} existing vulns")

    # Stage 5: PoC reproduction per verified vulnerability
    stage5_reports: list[str] = []
    if 5 not in config.skip_stages:
        if tui:
            tui.begin_stage(5, f"Reproducing {len(vuln_files)} vulnerabilities")
        stage5_reports = await run_stage5(vuln_files, config, checkpoint)
        if tui:
            tui.stage_progress(5, items_done=len(stage5_reports), items_total=len(vuln_files),
                               detail=f"{len(stage5_reports)} PoCs reproduced")
            tui.end_stage(5)
    else:
        logger.info("Stage 5 skipped. Loading existing reports.")
        stage5_dir = os.path.join(config.output_dir, "stage5-pocs")
        if os.path.isdir(stage5_dir):
            for name in sorted(os.listdir(stage5_dir), key=_natural_sort_key):
                entry = os.path.join(stage5_dir, name)
                if os.path.isdir(entry):
                    report = os.path.join(entry, "report.md")
                    if os.path.exists(report):
                        stage5_reports.append(report)
        if tui:
            tui.skip_stage(5, f"Loaded {len(stage5_reports)} existing reports")

    # Stage 6: disclosure preparation per reproduced vulnerability
    if 6 not in config.skip_stages:
        if tui:
            tui.begin_stage(6, f"Preparing {len(stage5_reports)} disclosures")
        await run_stage6(stage5_reports, config, checkpoint)
        if tui:
            tui.end_stage(6)
    else:
        logger.info("Stage 6 skipped.")
        if tui:
            tui.skip_stage(6)

    logger.info("Audit complete.")
